scalableminds/webknossos
# Connectome Tab

This PR implements the internal work package D.1 Connectome Viewing in webKnossos (see internal communication). Refer to the work package for the full feature set and a mockup.

This PR adds a new "Connectome Tab" and I tried to encapsulate the new functionality fairly well so that most of it is contained in a couple of files. Still, I had to adapt the code in a couple of spots to support some new use cases which I'll describe in the following:
- Allow to set the `enableRenderOnDemand` prop for layout tabs. Setting the prop to false will disable the lazy-loading of the tab and instead always render it. For the connectome tab, this is important to allow hotlinking to agglomerates/synapses without the tab being visible when loading the annotation. Otherwise, the code encapsulation would not work as well. I don't think this is a problem, though, since the tab doesn't do much if it's not actively being used.
- Allow to add multiple Skeletons (skeleton.js) to the scene. This is used to separate the connectome skeleton data from the "normal" skeleton data so the two don't interfere with each other and the connectome skeleton data is effectively read-only. Also allow to remove a Skeleton from the scene. This is done during cleanup or when the segmentation layer is switched.
- Introduce the `lastVisibleSegmentationLayerName` to allow to continue to use the connectome tab even if the segmentation was temporarily disabled. This might be useful in other spots as well.

### ToDo:
- [x] Separate connectome_view.js into separate logical units
- [x] Adapt docs/sharing.md#sharing-link-format to new link format
- [x] Check all check boxes that represent the same segment id (not only the clicked one). ~~No idea what would be a good way to do that yet~~
- [x] ~~Show the segment color next to the entry in the tree (if the agglomerate skeleton is loaded)~~ _Wait for feedback, maybe followup_
- [x] The connectome tab should work in view mode
- [x] Do we need a layout migration or is it sufficient to increase the layout version number?
- [x] Hovering over a Segment entry in the tab could highlight the segment in the data viewport. The same feature exists for the segment tab, so it might be easy to copy.
- [x] The context menu in the data viewport could have an entry for loading a segment into the connectome tab
- [x] If an agglomerate could not be loaded, because the node limit is exceeded, a nicer error message could be rendered. Currently, one has to expand the details of the error message to find out that this is not a bug.

### URL of deployed dev instance (used for testing):
- https://connectomeviewing.webknossos.xyz

### Steps to test:
- Open a dataset without a connectome. The connectome tab should state there is no connectome.
- Use the  `l4dense_motta_et_al_dev` dataset that has a linked connectome file on the deployed dev instance
- The connectome tab should work in view mode as well as in annotations
- Right-Click on a segment in the viewport and select "Import Agglomerate and Synapses". The segment should be shown in the connectome tab, its agglomerate skeleton loaded and all synapses shown as single node trees as well.
  - The same should happen if you enter the segment ID by hand
- Entering two segment IDs (comma-separated) should show both skeletons and only the synapses between the two agglomerates
- It should be possible to toggle the segmentation layer while a segment is shown in the connectome tab without the info in the connectome tab to disappear
- For datasets with multiple segmentation layers, if the segmentation layer is switched the tab should be reset
- The filters by synapse direction or synapse type should work and hide the filtered synapses/agglomerates from the tree list as well as the viewports
- Right-click on a 2nd-level segment in the tree list and select "Show All Synapses of This Segment" to see the synapses of that segment, instead
- The "Reset" buttons to reset the filters as well as the connectome tab should work
- It should be possible to link to a specific agglomerate and its synapses (tested in the screenshot tests), for example, [this link](https://connectomeviewing.webknossos.xyz/datasets/sample_organization/l4dense_motta_et_al_dev/view#{%22position%22:[2816,4543,1770],%22mode%22:%22orthogonal%22,%22zoomStep%22:0.734,%22stateByLayer%22:{%22segmentation%22:{%22connectomeInfo%22:{%22connectomeName%22:%22paper_l4_full_connectome%22,%22agglomerateIdsToImport%22:[1421,2717]}}}})

### Issues:
- fixes #5828 

------
